Prevent a OIDC panic if persister is nil

NMFR opened the pull request:
Using [oidc](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/plugin/pkg/client/auth/oidc/oidc.go) with a [DirectClientConfig](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/clientcmd/client_config.go#L77) without a [ConfigAccess](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/clientcmd/config.go#L34) (`nil`) would result in a panic due to the missing `persister` `nil` check.

I believe the correct behaviour is to proceed if no persister is specified. 
Another, less ideal, solution would be to return an error if no persister is present.

/kind bug

```release-note
fix OIDC auth provider panic if no persister is present
```

Fixes #90766

NMFR committed:
Prevent a OIDC panic if persister is nil

NMFR commented:
/kind bug
/sig cli

fedebongio commented:
/remove-sig api-machinery
/cc @mikedanese @enj

zhouya0 commented:
/ok-to-test

NMFR commented:
I don't see understand why / where the test is failing.
Could i get some help please?

Also should i remove the `// Persist new config and if successful, update the in memory config.` comment?
Seems to make less sense now and the code is pretty explicit on whats doing anyway.

NMFR committed:
Remove persister comment

Code is pretty explicit on whats doing.

NMFR commented:
Rebased from master and removed "Fixes #90766" from the commit message.

Also removed the `// Persist new config and if successful, update the in memory config.` comment.

NMFR commented:
/retest

NMFR commented:
/test pull-kubernetes-kubemark-e2e-gce-big

mikedanese commented:
Do we want to require a config persister or allow it to be nil?

NMFR commented:
In my opinion it should allow `nil`, there might be nothing to persist to. Having it `nil` instead of a no op persister is more explicit.

mikedanese commented:
It looks like gcp plugin handles this:

https://github.com/kubernetes/kubernetes/blob/9b44fa4a07b14b3dbc644f80e087d516677632f0/staging/src/k8s.io/client-go/plugin/pkg/client/auth/gcp/gcp.go#L224-L228

But the azure and (obviously) oidc plugin do not:

https://github.com/kubernetes/kubernetes/blob/9b44fa4a07b14b3dbc644f80e087d516677632f0/staging/src/k8s.io/client-go/plugin/pkg/client/auth/azure/azure.go#L344-L347

NMFR commented:
Shall i add the `nil` protection to azure too?

NMFR commented:
The gcp plugin is missing one `nil` protection too, in the `conditionalTransport.RoundTrip()`:

https://github.com/kubernetes/kubernetes/blob/9b44fa4a07b14b3dbc644f80e087d516677632f0/staging/src/k8s.io/client-go/plugin/pkg/client/auth/gcp/gcp.go#L375-L378

mikedanese commented:
Can you fix all of them and add a test? Noop persister might be the cleanest approach.

NMFR committed:
Add gcp auth provider nil persister test

NMFR committed:
Add azure auth provider nil persister test

NMFR committed:
Prevent gcp and azure auth plugin panic if persister is nil

NMFR commented:
Fixed all with if not `nil` logic, i prefer this strategy. 
But if you wish i can change and switch to a Noop persister.

Added tests to gcp and azure, skipped oidc because the tests there are very thin. If we need tests there i would suggest doing them in another issue due to complexity.

NMFR commented:
Can i get some help with the failing tests please?

I am not understanding how this changes are influencing the failing tests.

NMFR committed:
Update bazel go_default_test deps according with new dependency

NMFR commented:
rebase to master

NMFR commented:
@mikedanese can you help me understand the test errors please?

NMFR commented:
@enj @mikedanese could i get some helm interpreting the test errors please?

NMFR commented:
@enj @mikedanese could i get some help interpreting the test errors please?

NMFR commented:
@enj @mikedanese should i close this PR since it is going stale?
